src/resources/views/base/inc/alerts.blade.php

```diff
@@ -1,21 +1,40 @@
 {{-- Bootstrap Notifications using Prologue Alerts & PNotify JS --}}
 <script type="text/javascript">
-  Noty.overrideDefaults({
-    layout   : 'topRight',
-    theme    : 'backstrap',
-    timeout  : 2500, 
-    closeWith: ['click', 'button'],
-  });
 
-  @foreach (\Alert::getMessages() as $type => $messages)
+// This is intentionaly run after dom loads so this way we can avoid showing duplicate alerts
+// when the user is beeing redirected by persistent table, that happens before this event triggers.
 
-      @foreach ($messages as $message)
+window.addEventListener('DOMContentLoaded', function() {
 
-        new Noty({
-          type: "{{ $type }}",
-          text: "{!! str_replace('"', "'", $message) !!}"
-        }).show();
+    Noty.overrideDefaults({
+            layout   : 'topRight',
+            theme    : 'backstrap',
+            timeout  : 2500,
+            closeWith: ['click', 'button'],
+        });
 
-      @endforeach
-  @endforeach
-</script>
\ No newline at end of file
+    //get alerts from the alert bag
+    var $backpack_alerts = JSON.parse('@json(\Alert::getMessages())');
+
+    //the will be no simultaneous alerts, either we get them from the alert bag,
+    //or we already have them in the localstorage.
```

Comment:
I'm wondering. Is there any reason NOT to merge the arrays instead? We wouldn't use that in this particular scenario, but it would open the possibility of triggering notifications both from PHP and from JS using localstorage before redirect.

I believe this already impacts one often-overlooked notification: the "_You're using unlicensed software_" notification. That particular notification is triggered in PHP. If we overwrite the PHP array with the JS array (from localstorage) then that notification will no longer be shown.

But. If we _merged_ the two arrays... both notifications will be shown. So it'll still be annoying 🤣 

Of course this is one example, and I don't have another, but... personally I don't see why it wouldn't be better to merge the two arrays.

What do you think @pxpm ?

Reply:
Hey @tabacitu sorry it took so much time to have a look at this. Well, this is like the 5th time I give it look, but couldn't figure out what you mean. Help me here 👍 

I am not sure if I understood your question, but from my knowledge we don't merge it because when we add them to localstorage we already have ALL notifications that would be shown in that page, we just don't show in that page and show after redirection.

Like you said, if we merge both array we will have duplicated notifications, and we want to show only the notifications that would be shown to user if they haven't been redirected with persistent-table.

The only way I see this failing is when the notification is added after the datatables script run ? Is that possible ? JS runs on client machine, we already have all PHP notifications I think ?

Let me know what you think.

Best,
Pedro

Reply:
> The only way I see this failing is when the notification is added after the datatables script run ? Is that possible ? JS runs on client machine, we already have all PHP notifications I think ?

Exactly. And that's exactly what can happen with the "Unlicensed usage" notification, which randomly triggers a notification bubble - some page refreshes will have a bubble, some will not. 

So here's the scenario outlining the problem:
- upon pageload 1, Backpack will say "_Let's not show the unlicensed usage bubble_"; so it doesn't get put in `localStorage`;
- persistentTable refresh the page 
- upon pageload 2, Backpack will say "_I've had enough of this guy; show him an unlicensed usage bubble_" and trigger one (from PHP);
- but because the `localStorage` array overwrites the PHP array (currently), that bubble triggered from PHP will not get shown;

Again - this is a very niche use case. And I don't have another example to illustrate where this might happen. But I think it's an easy problem to fix, by just merging the two JSONs (even if most of the time the PHP one is empty). And I think it'll make for an easier code to read too.

I don't think notification duplication is an issue (I hope) - since notifications are send to page using flash data, the next request (upon page refresh) won't have them any more.

Also, what would make this easier to read (I think) is renaming the JS variables into `alerts_from_php` and `alerts_from_local_storage`.

Reply:
I get it now!!

Thanks for such good explanation. I will refactor accordingly.

Reply:
@tabacitu this wasn't just a matter of merging both. 

Before when we had alerts in `localStorage` was the indication that we had been redirected. When merging both of them we have no way to know when we should show the alerts or when we are about to be redirected. Also after redirection the PHP alerts were lost due the fact that they were only flashed to session for the next request.

To avoid this my solution was:
- In datatables ALLWAYS push the alerts to the localStorage and flush the AlertBag (this is done in `datatables_logic`)
- In the `alerts.blade.php` we merge the PHP alerts (probably the unlicenced one) with the LocalStorage and show them to the user.

I'v also updated the demo alerts: https://github.com/Laravel-Backpack/demo/pull/216

Let me know,
Pedro